src/raftstore/store/worker/region.rs

```diff
@@ -104,6 +105,38 @@ impl SnapContext {
         let raft_db = self.raft_db.clone();
         let raw_snap = Snapshot::new(self.kv_db.clone());
 
+        let mut old_snaps = None;
+        while self.mgr.get_total_snap_size() > self.mgr.max_total_snap_size() {
+            if old_snaps.is_none() {
+                let snaps = match self.mgr.list_idle_snap() {
+                    Ok(snaps) => snaps,
+                    Err(_) => break,
+                };
+                let mut key_and_snaps = Vec::with_capacity(snaps.len());
+                for (key, is_sending) in snaps {
+                    if !is_sending {
+                        continue;
+                    }
+                    let snap = match self.mgr.get_snapshot_for_sending(&key) {
+                        Ok(snap) => snap,
+                        Err(_) => continue,
+                    };
+                    if !snap.meta().and_then(|m| m.modified()).is_ok() {
+                        continue;
+                    }
+                    key_and_snaps.push((key, snap));
+                }
+                // Because we have filtered invalid `meta` and meta.modified`,
+                // we can unwrap safely here.
+                key_and_snaps.sort_by_key(|t| Reverse(t.1.meta().unwrap().modified().unwrap()));
+                old_snaps = Some(key_and_snaps);
+            }
+            match old_snaps.as_mut().unwrap().pop() {
+                Some((key, snap)) => self.mgr.delete_snapshot(&key, snap.as_ref(), false),
```

Comment:
should we ensure the snapshot can be deleted safely here? E.g, the snapshot is been sending.

Reply:
I think we don't need because we have covered the case in [here](https://github.com/pingcap/tikv/blob/master/src/server/snap.rs#L132)

Reply:
I think we don't need because we have covered the case in [here](https://github.com/pingcap/tikv/blob/master/src/server/snap.rs#L132)

Reply:
What if snapshot exists when check, but deleted when scanning?

Reply:
Good question. For current implement, I think it's ok because in function `get_snapshot_for_sending` we have read full meta into `snapshot`, and `snapshot.cf_files` have got file handles by call `File::open`. Other threads(or other processes of any user) can delete cf files safety as we have already got the file descriptor.